include/vcpkg/base/lazy.h

```diff
@@ -22,4 +25,25 @@ namespace vcpkg
         mutable T value;
         mutable bool initialized;
     };
+
+    template<class T>
+    struct AsyncLazy
+    {
+        template<class F>
+        AsyncLazy(F work) : value(std::async(std::launch::async | std::launch::deferred, [&work]() { return work(); }))
+        {
+        }
+
+        const T& get() const
+        {
+            if (std::holds_alternative<std::future<T>>(value))
+            {
+                value = std::get<std::future<T>>(value).get();
```

Comment:
This is out of contract, because you're trying to change which alternative in the variant is engaged while using a reference to the previously engaged element.

If simply reusing `future::get` is too expensive, that suggests you are trying to use futures too finely; they aren't cheap and `get` is the least of their expense.

Reply:
I'm not sure what you mean by that. The `std::future<T>` doesn't leave this function. The idea is to block until the result is available and store it as `T` because you can call `std::future.get()` only once. It's impossible to reuse `future.get()` as any additional call will result in `throw std::future_error{};`. I'd need `std::shared_future` for that but my intension was not to be thread safe here (`Lazy` isn't thread safe either).